File: frontend/src/services/reviewService.js
import { API_BASE_URL } from '../config/api';

export const reviewService = {
  async createReview(movieId, rating, context) {
    const token = localStorage.getItem('jwt');
    if (!token) {
      throw new Error('Bạn cần đăng nhập để đánh giá phim');
    }

    const response = await fetch(`${API_BASE_URL}/reviews`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`
      },
      body: JSON.stringify({
        movieId,
        rating,
        context: context || ''
      })
    });

    const data = await response.json();
    
    if (!response.ok || !data.success) {
      throw new Error(data.message || 'Có lỗi xảy ra khi gửi đánh giá');
    }

    return data;
  },

  async getReviewsByMovie(movieId) {
    const response = await fetch(`${API_BASE_URL}/reviews/movie/${movieId}`);
    const data = await response.json();
    
    if (!response.ok || !data.success) {
      throw new Error(data.message || 'Có lỗi xảy ra khi lấy đánh giá');
    }

    return data.data || [];
  },

  async getReviewsByUser(userId) {
    const token = localStorage.getItem('jwt');
    if (!token) {
      throw new Error('Bạn cần đăng nhập');
    }

    const response = await fetch(`${API_BASE_URL}/reviews/user/${userId}`, {
      headers: {
        'Authorization': `Bearer ${token}`
      }
    });
    
    const data = await response.json();
    
    if (!response.ok || !data.success) {
      throw new Error(data.message || 'Có lỗi xảy ra khi lấy đánh giá');
    }

    return data.data || [];
  },

  async reportReview(reviewId, reason) {
    const token = localStorage.getItem('jwt');
    if (!token) {
      throw new Error('Bạn cần đăng nhập để báo cáo đánh giá');
    }

    const response = await fetch(`${API_BASE_URL}/reviews/${reviewId}/report`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`
      },
      body: JSON.stringify({
        reason: reason || 'Nội dung không phù hợp'
      })
    });

    const data = await response.json();
    
    if (!response.ok || !data.success) {
      throw new Error(data.message || 'Có lỗi xảy ra khi báo cáo đánh giá');
    }

    return data;
  },

  // Admin methods
  async getAllReviews() {
    const token = localStorage.getItem('jwt');
    if (!token) {
      throw new Error('Bạn cần đăng nhập');
    }

    const response = await fetch(`${API_BASE_URL}/reviews/admin/all`, {
      headers: {
        'Authorization': `Bearer ${token}`
      }
    });
    
    const data = await response.json();
    
    if (!response.ok || !data.success) {
      throw new Error(data.message || 'Có lỗi xảy ra khi lấy danh sách đánh giá');
    }

    return data.data || [];
  },

  async getReportedReviews() {
    const token = localStorage.getItem('jwt');
    if (!token) {
      throw new Error('Bạn cần đăng nhập');
    }

    const response = await fetch(`${API_BASE_URL}/reviews/admin/reported`, {
      headers: {
        'Authorization': `Bearer ${token}`
      }
    });
    
    const data = await response.json();
    
    if (!response.ok || !data.success) {
      throw new Error(data.message || 'Có lỗi xảy ra khi lấy danh sách đánh giá bị báo cáo');
    }

    return data.data || [];
  },

  async toggleReviewVisibility(reviewId) {
    const token = localStorage.getItem('jwt');
    if (!token) {
      throw new Error('Bạn cần đăng nhập');
    }

    const response = await fetch(`${API_BASE_URL}/reviews/admin/${reviewId}/toggle-visibility`, {
      method: 'PUT',
      headers: {
        'Authorization': `Bearer ${token}`
      }
    });
    
    const data = await response.json();
    
    if (!response.ok || !data.success) {
      throw new Error(data.message || 'Có lỗi xảy ra khi thay đổi trạng thái đánh giá');
    }

    return data.data;
  },

  async getReviewReports(reviewId) {
    const token = localStorage.getItem('jwt');
    if (!token) {
      throw new Error('Bạn cần đăng nhập');
    }

    const response = await fetch(`${API_BASE_URL}/reviews/admin/${reviewId}/reports`, {
      headers: {
        'Authorization': `Bearer ${token}`
      }
    });
    
    const data = await response.json();
    
    if (!response.ok || !data.success) {
      throw new Error(data.message || 'Có lỗi xảy ra khi lấy danh sách báo cáo');
    }

    return data.data || [];
  },

  async deleteReview(reviewId) {
    const token = localStorage.getItem('jwt');
    if (!token) {
      throw new Error('Bạn cần đăng nhập');
    }

    const response = await fetch(`${API_BASE_URL}/reviews/${reviewId}`, {
      method: 'DELETE',
      headers: {
        'Authorization': `Bearer ${token}`
      }
    });
    
    const data = await response.json();
    
    if (!response.ok || !data.success) {
      throw new Error(data.message || 'Có lỗi xảy ra khi xóa đánh giá');
    }

    return data;
  }
};

File: backend/src/main/java/com/example/backend/services/ReviewService.java
package com.example.backend.services;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.example.backend.dtos.CreateReviewDTO;
import com.example.backend.dtos.ReviewResponseDTO;
import com.example.backend.entities.Movie;
import com.example.backend.entities.Review;
import com.example.backend.entities.ReviewReport;
import com.example.backend.entities.User;
import com.example.backend.repositories.MovieRepository;
import com.example.backend.repositories.ReviewRepository;
import com.example.backend.repositories.ReviewReportRepository;
import com.example.backend.repositories.UserRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class ReviewService {
    
    private final ReviewRepository reviewRepository;
    private final MovieRepository movieRepository;
    private final UserRepository userRepository;
    private final ReviewReportRepository reviewReportRepository;
    private final NotificationService notificationService;
    
    @Transactional
    public ReviewResponseDTO createReview(CreateReviewDTO createReviewDTO) {
        // Lấy username từ SecurityContext
        String username = SecurityContextHolder.getContext().getAuthentication().getName();
        
        // Tìm user
        Optional<User> userOpt = userRepository.findByUsername(username);
        if (userOpt.isEmpty()) {
            throw new RuntimeException("Không tìm thấy người dùng");
        }
        User user = userOpt.get();
        
        // Tìm movie
        Optional<Movie> movieOpt = movieRepository.findById(createReviewDTO.getMovieId());
        if (movieOpt.isEmpty()) {
            throw new RuntimeException("Không tìm thấy phim");
        }
        Movie movie = movieOpt.get();
        
        // Kiểm tra xem user đã đánh giá phim này chưa
        Optional<Review> existingReview = reviewRepository.findByUserUserIdAndMovieMovieId(
            user.getUserId(), 
            createReviewDTO.getMovieId()
        );
        
        Review review;
        if (existingReview.isPresent()) {
            // Cập nhật đánh giá đã có
            review = existingReview.get();
            review.setRating(createReviewDTO.getRating());
            review.setContext(createReviewDTO.getContext());
            review.setCreatedUpdate(LocalDateTime.now());
        } else {
            // Tạo đánh giá mới
            review = Review.builder()
                .user(user)
                .movie(movie)
                .rating(createReviewDTO.getRating())
                .context(createReviewDTO.getContext())
                .createdAt(LocalDateTime.now())
                .createdUpdate(LocalDateTime.now())
                .build();
        }
        
        Review savedReview = reviewRepository.save(review);
        
        // Gửi thông báo WebSocket khi đánh giá thành công
        notificationService.notifyReviewSuccess(user.getUserId(), movie.getTitle());
        
        return mapToDTO(savedReview);
    }
    
    public List<ReviewResponseDTO> getReviewsByMovie(Long movieId) {
        // Only return reviews that are not hidden
        List<Review> reviews = reviewRepository.findByMovieMovieIdAndIsHiddenFalse(movieId);
        return reviews.stream()
            .map(this::mapToDTO)
            .collect(Collectors.toList());
    }
    
    public List<ReviewResponseDTO> getReviewsByUser(Long userId) {
        List<Review> reviews = reviewRepository.findByUserUserId(userId);
        return reviews.stream()
            .map(this::mapToDTO)
            .collect(Collectors.toList());
    }
    
    @Transactional
    public void reportReview(Long reviewId, String reason) {
        // Lấy username từ SecurityContext
        String username = SecurityContextHolder.getContext().getAuthentication().getName();
        
        // Tìm user
        Optional<User> userOpt = userRepository.findByUsername(username);
        if (userOpt.isEmpty()) {
            throw new RuntimeException("Không tìm thấy người dùng");
        }
        User user = userOpt.get();
        
        // Tìm review
        Optional<Review> reviewOpt = reviewRepository.findById(reviewId);
        if (reviewOpt.isEmpty()) {
            throw new RuntimeException("Không tìm thấy đánh giá");
        }
        Review review = reviewOpt.get();
        
        // Kiểm tra xem user đã report review này chưa
        Optional<ReviewReport> existingReport = reviewReportRepository.findByReviewReviewIdAndUserUserId(
            reviewId, 
            user.getUserId()
        );
        
        if (existingReport.isPresent()) {
            throw new RuntimeException("Bạn đã báo cáo đánh giá này rồi");
        }
        
        // Tạo report mới
        ReviewReport report = ReviewReport.builder()
            .review(review)
            .user(user)
            .reason(reason)
            .build();
        
        reviewReportRepository.save(report);
        
        // Tăng reportCount
        review.setReportCount(review.getReportCount() + 1);
        reviewRepository.save(review);
    }
    
    public List<ReviewResponseDTO> getAllReviews() {
        // Get all reviews for admin management
        List<Review> reviews = reviewRepository.findAllByOrderByCreatedAtDesc();
        return reviews.stream()
            .map(this::mapToDTO)
            .collect(Collectors.toList());
    }
    
    public List<ReviewResponseDTO> getReportedReviews() {
        // Get all reviews that have been reported (reportCount > 0)
        List<Review> reviews = reviewRepository.findByReportCountGreaterThanOrderByReportCountDesc(0);
        return reviews.stream()
            .map(this::mapToDTO)
            .collect(Collectors.toList());
    }
    
    @Transactional
    public ReviewResponseDTO toggleReviewVisibility(Long reviewId) {
        Optional<Review> reviewOpt = reviewRepository.findById(reviewId);
        if (reviewOpt.isEmpty()) {
            throw new RuntimeException("Không tìm thấy đánh giá");
        }
        
        Review review = reviewOpt.get();
        review.setIsHidden(!review.getIsHidden());
        Review savedReview = reviewRepository.save(review);
        
        return mapToDTO(savedReview);
    }
    
    private ReviewResponseDTO mapToDTO(Review review) {
        return ReviewResponseDTO.builder()
            .reviewId(review.getReviewId())
            .userId(review.getUser().getUserId())
            .username(review.getUser().getUsername())
            .movieId(review.getMovie().getMovieId())
            .movieTitle(review.getMovie().getTitle())
            .rating(review.getRating())
            .context(review.getContext())
            .createdAt(review.getCreatedAt())
            .updatedAt(review.getCreatedUpdate())
            .isHidden(review.getIsHidden())
            .reportCount(review.getReportCount())
            .build();
    }
}


